Rewrite of `SuperchainWETH` (#1297)
ethereum-optimism/docs: pages/stack/interop/superchain-weth.mdx

---
title: Interoperable ETH
lang: en-US
description: Learn basic details about Interoperable ETH.
---

import { Callout } from 'nextra/components'

import { InteropCallout } from '@/components/WipCallout'

<InteropCallout />

# Interoperable ETH

InteroperableETH is implemented using three contracts:

- [`SuperchainWETH`](https://github.com/ethereum-optimism/optimism/blob/develop/contracts-bedrock/src/L2/SuperchainWETH.sol) is the bridge that lets you move ETH from one Superchain blockchain to another.

> - [`ETHLiquidity`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/ETHLiquidity.sol) holds a practically infinite amount of ETH (it starts with 2<sup>248</sup> WEI).

Reviewer: think we can nix “holds a practically infinite amount…”

    }
  }
}%%

sequenceDiagram
  box rgba(0,0,0,0.1) Source Chain
    participant src-liquidity as ETHLiquidity
    participant src-weth as SuperchainWETH
  end
  actor user as User
  participant any-offchain as User, Relayer, etc.  
  box rgba(0,0,0,0.1) Destination Chain
    participant dst-l2Xl2 as L2ToL2CrossDomainMessenger
    participant dst-weth as SuperchainWETH
    participant dst-liquidity as ETHLiquidity       
    participant dst-final as ETH Destination
  end
  rect rgba(0,0,0,0.1)
    note over src-weth, dst-l2Xl2: Initiating Message
    user->>src-weth: 1. sendETH(to, chainId) with n ETH
    src-weth->>src-liquidity: 2. Deposit n ETH
    src-weth->>dst-l2Xl2: 3. Send relayETH to Superchain WETH on chainId
  end
  rect rgba(0,0,0,0.1)
    note over any-offchain,dst-weth: Executing message
    any-offchain->>dst-l2Xl2: 4. Relay the message
    dst-l2Xl2->>dst-weth: 5. relayETH(from, to, amount)
    dst-weth->>dst-liquidity: 6. Withdraw amount ETH
    dst-weth->>dst-final: 7. Here are amount ETH
  end
```

#### Initiating message

1.  The user (or a contract operating on a user's behalf) calls `SuperchainWETH.sendETH` with a destination address and a chainId.

>     This call should have an attached amount of ETH.

Author: This gives the wrong impression, IMAO, that the call includes the amount of ETH to transfer as a parameter, rather than that the ETH is moved with the call. 

See what I wrote instead.

2.  `SuperchainWETH` sends the amount of ETH to `ETHLiquidity`, removing it from circulation on the source chain.

3.  `SuperchainWETH` on the source chain sends a [`relayETH`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/SuperchainWETH.sol#L125-L145) call to the `SuperchainWETH` on the destination chain using [`L2ToL2CrossDomainMessenger`](./message-passing).

#### Executing message

> 4.  Some offchain entity submits a transaction with the executing message.
>     Any address can submit such a transaction, but it has to have ETH on the destination chain to do so.
>     We expect that usually this will be a relayer rather than the user because the user won't have ETH on the destination chain yet.

Reviewer 2: ```suggestion
4.  An off-chain entity submits a transaction to execute the message.
    Any address can submit this transaction, but it must have ETH on the destination chain.
    Typically, a relayer submits the transaction, since the user does not yet have ETH on the destination chain.
```

> 5.  `L2ToL2CrossDomainMessenger` on the destination chain calls `SuperchainWETH` with the source of the ETH, the destination address, and the amount.

Reviewer 2: ```suggestion
5.  `L2ToL2CrossDomainMessenger` on the destination chain calls `SuperchainWETH` with the following details:

*   Source of the ETH
*   Destination address
*   Amount of ETH
```

>     `SuperchainWETH` runs several sanity checks:
> 
>     *   The `relayETH` call has to come directly from `L2ToL2CrossDomainMessenger`.
>     *   The interop message has to have been sent by `SuperchainWETH`

Reviewer 2: ```suggestion
    `SuperchainWETH` performs several sanity checks:

    *   The `relayETH` call must originate from `L2ToL2CrossDomainMessenger`.
    *   The interop message must have been sent by `SuperchainWETH`
```

> 6.  `SuperchainWETH` withdraws the requested amount of ETH from `ETHLiquidity`.
>     It is the only contract allowed to withdraw from `ETHLiquidity`, which adds to the ETH in circulation on the destination chain.

Reviewer 2: ```suggestion
6.  `SuperchainWETH` withdraws the specified amount of ETH from `ETHLiquidity`.
    Only `SuperchainWETH` can withdraw from `ETHLiquidity`, ensuring that the ETH is correctly reintroduced into circulation on the destination chain.
```

> 7.  `SuperchainWETH` uses [`SafeSend`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/SafeSend.sol) to send the ETH.

Reviewer 2: ```suggestion
7.  `SuperchainWETH` uses [`SafeSend`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/SafeSend.sol) to send ETH.
```

>     This means that even if the destination is a smart contract, its custom logic does not get called, [in contrast to normal ETH transfers](https://docs.base.org/base-learn/docs/address-and-payable/#receiving-ether-with-payable-addresses).

Reviewer 2: ```suggestion
    This ensures that even if the destination is a smart contract, its custom logic is not executed.
This behavior differs from [standard ETH transfers](https://docs.base.org/base-learn/docs/address-and-payable/#receiving-ether-with-payable-addresses), where smart contracts can trigger custom logic upon receiving ETH.
```

## L1 Treasury

> Every ETH in circulation on the Superchain (all the ETH except for those held by `ETHLiquidity`) needs to be backed by ETH held on L1.
> This is done using a lockbox contract on L1 that holds all the ETH ever bridged to [chains in the Superchain interop cluster](explainer#superchain-interop-cluster) and not yet withdrawn.

Reviewer 2: ```suggestion
Every ETH in circulation on the Superchain—excluding ETH held by `ETHLiquidity`—must be backed by ETH on L1.

This is enforced by a lockbox contract on L1, which holds all ETH bridged to [Superchain interop cluster chains](https://github.com/ethereum-optimism/docs/pull/1297/explainer#superchain-interop-cluster) that has not yet been withdrawn.
```

    The user sends 3 ETH on L1 to the bridge, which is locked in the lockbox.
    The bridge on chain A then mints 3 ETH for the user.

3.  The user sent the initiating message to `SuperchainWETH` on chain A, along with 1 ETH to bridge to chain B.
    This 1 ETH is sent to `ETHLiquidity` on chain A.

4.  Somebody (the user, a relayer action on behalf of the user, etc.) sent the corresponding executing message to chain B.
    `SuperchainWETH` transfers 1 ETH from `ETHLiquidity` on chain B to the user.

5.  The user decides to withdraw 1 ETH from chain B back into L1.
    Normally, a user would do this through a third-party bridge, which is faster and usually cheaper, but for illustration purposes this user uses the standard OP bridge.
    The user starts with an initiating message on chain B, which burns 1 ETH and sends a message to L1.

6.  After the week long [challenge period](/connect/resources/glossary#challenge-period), the user finalizes the withdrawal on L1.
    The lock box releases 1 ETH, which is then sent to the user.

## Next steps

*   Explore [the `SuperchainWETH` specs](https://specs.optimism.io/interop/superchain-weth.html) for in-depth implementation details.

> *   Read the [interop message passing](./message-passing) page for more information about how `L2ToL2CrossDomainMessenger` passes messages.

Reviewer 2: ```suggestion
*   Read the [interop message passing](/message-passing) page for more information about how `L2ToL2CrossDomainMessenger` passes messages.
```

Reviewer 2: It works by default on nextra.

    We expect that in many cases this will be a relayer rather than the user, because the user won't have ETH on the destination chain yet.

5.  `L2ToL2CrossDomainMessenger` on the destination chain calls `SuperchainWETH` with the source of the ETH, the destination address, and the amount.

    `SuperchainWETH` runs several sanity checks:

    *   The `relayETH` call has to come directly from `L2ToL2CrossDomainMessenger`.
    *   The interop message has to have been sent by `SuperchainWETH`

6.  `SuperchainWETH` withdraws the requested amount of ETH from `ETHLiquidity`.
    It is the only contract allowed to withdraw from `ETHLiquidity`, which adds to the ETH in circulation on the destination chain.

7.  `SuperchainWETH` uses [`SafeSend`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/SafeSend.sol) to send the ETH.
    This means that even if the destination is a smart contract, its custom logic does not get called, [in contrast to normal ETH transfers](https://docs.base.org/base-learn/docs/address-and-payable/#receiving-ether-with-payable-addresses).

## L1 Treasury

> All the ETH in circulation on the Superchain (all the ETH except for what is held by `ETHLiquidity`) needs to be backed by real ETH held on L1.

Reviewer: needs to be backed by ETH held on L1.

(remove word “real”)

> This is done using a lockbox contract on L1 that holds all the ETH ever bridged to Superchain chains and not yet withdrawn.

Reviewer: ETH ever bridged to chains in the Superchain interop cluster

(consider linking back to cluster def in explainer)

*   Source of the ETH
*   Destination address
*   Amount of ETH

    `SuperchainWETH` performs several sanity checks:

    *   The `relayETH` call must originate from `L2ToL2CrossDomainMessenger`.
    *   The interop message must have been sent by `SuperchainWETH`

6.  `SuperchainWETH` withdraws the specified amount of ETH from `ETHLiquidity`.
    Only `SuperchainWETH` can withdraw from `ETHLiquidity`, ensuring that the ETH is correctly reintroduced into circulation on the destination chain.

7.  `SuperchainWETH` uses [`SafeSend`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/SafeSend.sol) to send ETH.
    This ensures that even if the destination is a smart contract, its custom logic is not executed.
This behavior differs from [standard ETH transfers](https://docs.base.org/base-learn/docs/address-and-payable/#receiving-ether-with-payable-addresses), where smart contracts can trigger custom logic upon receiving ETH.

## L1 Treasury

Every ETH in circulation on the Superchain—excluding ETH held by `ETHLiquidity`—must be backed by ETH on L1.

> This is enforced by a lockbox contract on L1, which holds all ETH bridged to [Superchain interop cluster chains](https://github.com/ethereum-optimism/docs/pull/1297/explainer#superchain-interop-cluster) that has not yet been withdrawn.

Reviewer 2: ```suggestion
This is enforced by a lockbox contract on L1, which holds all ETH bridged to [Superchain interop cluster chains](/explainer#superchain-interop-cluster) that has not yet been withdrawn.
```
